projects/ngx-fetcher-with-etag/src/lib/fetcher-with-etag.ts

```diff
@@ -0,0 +1,275 @@
+    if (isValidRefreshInterval) {
+      const refreshInterval = this.lastFetchTimestamp! + this.refreshInterval - now;
+      if (interval === 0 || interval > refreshInterval) {
+        interval = refreshInterval;
+      }
+    }
+
+    const isInvalidInterval =
+      interval === 0 && this.value !== undefined && !this.expires && !this.refreshInterval;
+    if (isInvalidInterval) {
+      return undefined;
+    }
+    return interval;
+  }
+
+  /**
+   * Fetches the value if needed, otherwise returns the old value.
+   *
+   * If a fetch was executed before and the known value is fresh, then it's returned without further network access.
+   *
+   * The value is considered fresh if `Expires` is in the future or if the last fetch timestamp is within the
+   * refresh interval.
+   *
+   * If there was an `ETag` header in the previous response, it's sent as `If-None-Match` and if the server
+   * replies 304 (Not Modified) then the previous value is reused.
+   *
+   * If the server returns a new value (200), then the value is converted using `converter`.
+   */
+  async fetch(): Promise<T> {
+    const interval = this.getRefreshIntervalInMilliseconds();
+    if (interval === undefined) {
+      if (this.fetchRefreshesWithoutInterval) {
+        console.debug(`${this.url}: interval=undefined && fetchRefreshesWithoutInterval=true`);
+      } else {
+        console.debug(`${this.url}: is never to be refreshed`);
+        return this.value!;
+      }
+    }
+    if ((interval ?? 0) > 0) {
+      console.debug(`${this.url}: still fresh for ${interval} milliseconds`);
+      return this.value!;
+    }
+    if (!this.url) {
+      throw new Error('url is not set');
+    }
+
+    const headers = this.getHeaders();
+    this.loadingSubject.next(true);
+    const response = await lastValueFrom(
+      this.httpClient.get<RawT>(this.url, { headers, observe: 'response' })
+    );
+    this.loadingSubject.next(false);
+
+    if (response.status !== HttpStatusCode.Ok && response.status !== HttpStatusCode.NotModified) {
+      console.error(`${this.url}: unexpected response`, response);
+      throw new Error(`unexpected response ${response.status} for url ${this.url}`);
+    }
+
+    this.etag = response.headers.get('ETag');
+    this.lastFetchTimestamp = Date.now();
+    this.expires = this.createExpires(response.headers.get('Expires'));
+
+    if (response.status === HttpStatusCode.NotModified) {
+      console.log(`${this.url}: response 304 (Not Modified) - reusing data`, response);
+    } else if (!response.body) {
+      console.warn(`${this.url}: response 200 without a body - ignored`, response);
+    } else {
+      const newValue = this.converter(response.body);
+      if (this.value === undefined || !this.isEqual(this.value, newValue)) {
+        this.value = newValue;
+        this.dataSubject.next(this.value!);
+      } else {
+        console.log(`${this.url}: response 200 with equal value - ignored`);
+      }
+    }
+    return this.value!;
```

Comment:
very small nit: I don't think this `else if` is needed. Wouldn't something like this work the same way? IMO, in the `else if` you are already dealing with the response status `200` so it would make more sense to keep it as an `else`. But that's your call
```suggestion
    if (response.status === HttpStatusCode.NotModified) {
      console.log(`${this.url}: response 304 (Not Modified) - reusing data`, response);
    } else {
        if (!response.body) {
          console.warn(`${this.url}: response 200 without a body - ignored`, response);
        } else {
          const newValue = this.converter(response.body);
          if (this.value === undefined || !this.isEqual(this.value, newValue)) {
            this.value = newValue;
            this.dataSubject.next(this.value!);
          } else {
            console.log(`${this.url}: response 200 with equal value - ignored`);
          }
        }
      }
      return this.value!;
    }
```

Reply:
It makes sense, but I think that is related with the @WilsonNet [comment](https://github.com/profusion/pf-angular-extensions/pull/1#discussion_r2334341334) about nest code. Even though they both work the same way, the second one is the harder one to review